src/lib/CoreInstance.cc

```diff
@@ -0,0 +1,304 @@
+  } else if (type == simeng::MemInterfaceType::Fixed) {
+    instructionMemory_ = std::make_shared<simeng::FixedLatencyMemoryInterface>(
+        processMemory_.get(), processMemorySize_,
+        config_["LSQ-L1-Interface"]["Access-Latency"].as<uint16_t>());
+  } else {
+    std::cerr << "Unsupported memory interface type used in "
+                 "createL1InstructionMemory()."
+              << std::endl;
+    exit(1);
+  }
+
+  return;
+}
+
+void CoreInstance::setL1InstructionMemory(
+    std::shared_ptr<simeng::MemoryInterface> memRef) {
+  // Set the L1I cache instance to use
+  instructionMemory_ = memRef;
+  return;
+}
+
+void CoreInstance::createL1DataMemory(const simeng::MemInterfaceType type) {
+  // Create a L1D cache instance based on type supplied
+  if (type == simeng::MemInterfaceType::Flat) {
+    dataMemory_ = std::make_shared<simeng::FlatMemoryInterface>(
+        processMemory_.get(), processMemorySize_);
+  } else if (type == simeng::MemInterfaceType::Fixed) {
+    dataMemory_ = std::make_shared<simeng::FixedLatencyMemoryInterface>(
+        processMemory_.get(), processMemorySize_,
+        config_["LSQ-L1-Interface"]["Access-Latency"].as<uint16_t>());
+  } else {
+    std::cerr
+        << "Unsupported memory interface type used in createL1DataMemory()."
+        << std::endl;
+    exit(1);
+  }
+
+  return;
+}
+
+void CoreInstance::setL1DataMemory(
+    std::shared_ptr<simeng::MemoryInterface> memRef) {
+  // Set the L1D cache instance to use
+  dataMemory_ = memRef;
+  return;
+}
+
+void CoreInstance::createCore() {
+  // If memory interfaces must be manually set, ensure they have been
+  if (setDataMemory_ && dataMemory_ == nullptr) {
+    std::cerr << "Data memory not set. External Data memory must be manually "
+                 "set using the setL1DataMemory(...) function."
+              << std::endl;
+    exit(1);
+  } else if (setInstructionMemory_ && instructionMemory_ == nullptr) {
+    std::cerr << "Instruction memory not set. External instruction memory "
+                 "interface must be manually set using the "
+                 "setL1InstructionMemory(...) function."
+              << std::endl;
+    exit(1);
+  }
+
+  // Construct architecture object
+  arch_ =
+      std::make_unique<simeng::arch::aarch64::Architecture>(kernel_, config_);
+
+  // Construct branch predictor object
+  predictor_ = std::make_unique<simeng::GenericPredictor>(config_);
+
+  // Extract port arrangement from config file
+  auto config_ports = config_["Ports"];
+  std::vector<std::vector<uint16_t>> portArrangement(config_ports.size());
+  for (size_t i = 0; i < config_ports.size(); i++) {
+    auto config_groups = config_ports[i]["Instruction-Group-Support"];
+    // Read groups in associated port
+    for (size_t j = 0; j < config_groups.size(); j++) {
+      portArrangement[i].push_back(config_groups[j].as<uint16_t>());
+    }
+  }
+  portAllocator_ = std::make_unique<simeng::pipeline::BalancedPortAllocator>(
```

Comment:
Given there are multiple port allocator classes available, could a config option be introduced to define which one is used?

Reply:
Similar to https://github.com/UoB-HPC/SimEng/pull/247#discussion_r972132103, good idea but out of scope.